Allow using and registering skyboxes in code

BoogieMonster1O1 opened the pull request:
This pr basically does three things
- Allows adding permanent skyboxes that are stored in a list that is not cleared during a resource reload.
- Replace suppliers that were used to get codecs and types with `SkyboxType` and a registry.
- Replace legacy deserialization code inside skyboxes with separate deserializers. Also uses a registry.

I've also added fabric api as a dependency instead of just resource loader so that I can use registry sync's registry registering hooks

BoogieMonster1O1 committed:
Actually depend on fabric api

 Changes to be committed:
	modified:   build.gradle
	modified:   gradle.properties

BoogieMonster1O1 committed:
Registry

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java
	new file:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java
	new file:   src/main/java/io/github/amerebagatelle/fabricskyboxes/util/object/internal/Metadata.java

BoogieMonster1O1 committed:
Yeet parsers

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/resource/SkyboxResourceListener.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/AbstractSkybox.java
	new file:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/LegacyDeserializer.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/MonoColorSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/AnimatedSquareTexturedSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/SquareTexturedSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/TexturedSkybox.java

BoogieMonster1O1 committed:
It compiles

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/resource/SkyboxResourceListener.java

BoogieMonster1O1 committed:
Yeet getCodec

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/FabricSkyBoxesClient.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/AbstractSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/MonoColorSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/AnimatedSquareTexturedSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/SquareTexturedSkybox.java

BoogieMonster1O1 committed:
Test mod

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/AbstractSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/MonoColorSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/AnimatedSquareTexturedSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/SquareTexturedSkybox.java
	new file:   src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java
	new file:   src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java
	modified:   src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Test.java

BoogieMonster1O1 committed:
Actually deserialize legacy skyboxes

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/resource/SkyboxResourceListener.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/LegacyDeserializer.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/AnimatedSquareTexturedSkybox.java

BoogieMonster1O1 committed:
Build registries with MODDED attribute

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/LegacyDeserializer.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java

BoogieMonster1O1 committed:
Remove include

 Changes to be committed:
	modified:   build.gradle

AMereBagatelle commented on build.gradle:
```diff
@@ -31,8 +31,8 @@ dependencies {
-    modCompile "net.fabricmc.fabric-api:fabric-resource-loader-v0:${project.fabric_resource_loader_version}"
-    include "net.fabricmc.fabric-api:fabric-resource-loader-v0:${project.fabric_resource_loader_version}"
+    modCompile "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
+    include "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
```
Yeah, no.  I personally don't like the idea of including all of FAPI.  What other parts of it do you use?

FlashyReese replied:
It's implementing only

BoogieMonster1O1 replied:
to answer your question: resource loader and registry sync
jijing registry sync is concern and its something nobody ever does

BoogieMonster1O1 committed:
Imports

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/FabricSkyBoxesClient.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/AbstractSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/LegacyDeserializer.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/MonoColorSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/SkyboxType.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/AnimatedSquareTexturedSkybox.java
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/textured/SquareTexturedSkybox.java

BoogieMonster1O1 committed:
Fix an issue with static initializers running too early

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/skyboxes/LegacyDeserializer.java
	modified:   src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Test.java
	modified:   src/testmod/resources/fabric.mod.json

BoogieMonster1O1 commented:
The Test mod works. Now ready for review

AMereBagatelle commented on src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java:
```diff
@@ -22,49 +22,36 @@
-    private static final ArrayList<AbstractSkybox> skyboxes = new ArrayList<>();
+    private final Predicate<? super AbstractSkybox> renderPredicate = (skybox) -> !this.activeSkyboxes.contains(skybox) && skybox.alpha >= 0.1;
+    private final ArrayList<AbstractSkybox> skyboxes = new ArrayList<>();
+    private final ArrayList<AbstractSkybox> permanentSkyboxes = new ArrayList<>();
```
I personally require a little more description on all of this, seems a little unclear what permanentSkyboxes actually does for us

AMereBagatelle commented on src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java:
```diff
@@ -0,0 +1,24 @@
+import io.github.amerebagatelle.fabricskyboxes.util.object.DefaultProperties;
+import io.github.amerebagatelle.fabricskyboxes.util.object.RGBA;
+
+public class AnEntirelyHardcodedSkybox extends MonoColorSkybox {
```
Change the name, something like "HardcodedTestSkybox" would be better.

BoogieMonster1O1 replied:
this is the testmod

AMereBagatelle replied:
I know, and that's why.  I don't feel like `AnEntirelyHardcodedSkybox` really is good code style naming

BoogieMonster1O1 replied:
the testmod is mostly implementation detail, but sure ill rename it

AMereBagatelle commented on src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java:
```diff
@@ -0,0 +1,17 @@
+
+import net.fabricmc.api.ClientModInitializer;
+
+public class Init implements ClientModInitializer {
```
Change name to something like "FSTestInit"

BoogieMonster1O1 replied:
:eyes:
sure

BoogieMonster1O1 replied:
but its the testmod

AMereBagatelle replied:
I know it's picky... it's a personal code style thing :P

AMereBagatelle reviewed (changes requested):
Looks good in general, still have to test it locally but here's the things I found in code

BoogieMonster1O1 committed:
Rename testmod classes

 Changes to be committed:
	renamed:    src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java -> src/testmod/java/io/github/amerebagatelle/fabricskyboxes/TestClientModInitializer.java
	renamed:    src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Test.java -> src/testmod/java/io/github/amerebagatelle/fabricskyboxes/TestPreLaunchEntrypoint.java
	renamed:    src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java -> src/testmod/java/io/github/amerebagatelle/fabricskyboxes/TestSkybox.java
	modified:   src/testmod/resources/fabric.mod.json

BoogieMonster1O1 committed:
Document and annotate skybox manager members

 Changes to be committed:
	modified:   src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java

BoogieMonster1O1 committed:
Api documentation

 Changes to be committed:
	new file:   docs/api.md

BoogieMonster1O1 commented:
Renamed testmod classes and added javadoc. 

I also annotated some of the methods in `SkyboxManager` to specify that they're not part of the public api (they should also probably be moved to a separate class (`SkyboxManagerImpl`?) in a later version)

AMereBagatelle commented:
I was considering doing something of the sort to SkyboxManager, also I think this is now ready for merge

AMereBagatelle commented on docs/api.md:
```diff
@@ -0,0 +1,9 @@
+# Api Documentation
+### Permanent skyboxes
+FabricSkyBoxes provides an api that allows you to register skyboxes in code. 
+These skyboxes will never be cleared after a resource reload and should often only be used to provide custom sky rendering to your custom dimension.
```
Or biome :D

BoogieMonster1O1 replied:
fixed

BoogieMonster1O1 committed:
Update api.md
